Docs for data stream REST APIs

danhermann opened the pull request:
Adds documentation for the basic create, get, and delete REST APIs for data streams.

Relates to #53100

danhermann committed:
Docs for data stream REST APIs

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

danhermann committed:
add to index APIs page

martijnvg reviewed (approved):
This looks great! LGTM (when build succeeds)

As a followup we should also replace add a general data stream page (which replaces the redirect in `redirects.asciidoc` around line 390).

danhermann committed:
update redirects

danhermann committed:
update section name

jrodewig commented on docs/reference/indices/delete-data-stream.asciidoc:
```diff
@@ -0,0 +1,29 @@
+[[delete-data-stream-api-path-params]]
+==== {api-path-parms-title}
+
+`<data-stream>`::
++
+--
+(Required, string) Name or wildcard expression of data stream(s) to delete.
```
This is created an unterminated block, which is breaking the docs build. Since the definition is a single para, we don't need the `+` or `--` (block).
```suggestion
`<data-stream>`::
(Required, string) Name or wildcard expression of data stream(s) to delete.
```

danhermann committed:
Update docs/reference/indices/delete-data-stream.asciidoc

Co-Authored-By: James Rodewig <james.rodewig@elastic.co>

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

danhermann committed:
fix test directives

danhermann committed:
more test directives

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

henningandersen commented on docs/reference/indices/create-data-stream.asciidoc:
```diff
@@ -0,0 +1,73 @@
+attribute that indicates which of the backing indices is the write index, the index into which documents
+will be ingested.
+
+Backing indices are always named `<data-stream-name>-zzzzzz` where `zzzzzz` is the six-digit, zero-padded
```
I think we will violate this naming convention once we can shrink an index in a data stream (shrink in ilm creates a new index named `"shrink-<index-name>"` and since we cannot rename, we are likely stuck with that name. 

So we should loosen the wording here a bit (always at least seems too strong).

henningandersen commented on docs/reference/indices/create-data-stream.asciidoc:
```diff
@@ -0,0 +1,73 @@
+Data stream names must meet the following criteria:
+
+- Lowercase only
+- Cannot include `\`, `/`, `*`, `?`, `"`, `<`, `>`, `|`, ` ` (space character), `,`, `#`
```
I think we should also disallow ':' as for indices?

henningandersen commented on docs/reference/indices/get-data-stream.asciidoc:
```diff
@@ -0,0 +1,86 @@
+////
+[source,console]
+-----------------------------------
+DELETE /_data_stream/foo
```
I think this should delete `my-data-stream` rather than `foo`?

henningandersen commented on docs/reference/indices/get-data-stream.asciidoc:
```diff
@@ -0,0 +1,86 @@
+[[get-data-stream-api-request]]
+==== {api-request-title}
+
+`GET _data_streams/<data-stream>`
```
I think this should be singular:

```suggestion
`GET _data_stream/<data-stream>`
```

danhermann replied:
It is `_data_streams` (plural) which I see is inconsistent with many of the other endpoints such as `_alias`, `_index_template`, `_security/role`, etc., that use the singular for retrieving information about one or more resources. @martijnvg, do you think we should change this?

martijnvg replied:
Yes, I think we should change this to `_data_stream` in order to make this consistent with other apis.

danhermann replied:
Fyi, we will change this endpoint to use the singular `_data_stream` in a series of follow-up PRs.

henningandersen commented on docs/reference/indices/get-data-stream.asciidoc:
```diff
@@ -0,0 +1,86 @@
+
+[source,console]
+--------------------------------------------------
+GET _data_streams/my-data-stream*
```
Also singular I think?

```suggestion
GET _data_stream/my-data-stream*
```

henningandersen reviewed (approved):
LGTM.

danhermann committed:
review comments, wrap columns at 80

danhermann committed:
Merge branch 'docs_data_stream_crud_apis' of https://github.com/danhermann/elasticsearch into docs_data_stream_crud_apis

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs_data_stream_crud_apis

martijnvg committed:
Merge remote-tracking branch 'es/master' into docs_data_stream_crud_apis

martijnvg committed:
Turned separate snippets into a single 'suite'.
This fixed test snippet in create-data-stream.asciidoc,
because after the first snippet the tear down was invoked and
deleting all indices failed, because there was a data stream.

martijnvg committed:
Disable shard failures check, if response has top level array then it can't figure out if the response had shard failures.
Disabled response assertions as we can't seem to string replace if response has top level array.
